packages/byods/src/TokenManager.ts

```diff
@@ -0,0 +1,187 @@
+      orgId,
+      token: {
+        accessToken: data.access_token,
+        refreshToken: data.refresh_token,
+        expiresAt: new Date(Date.now() + data.expires_in * 1000),
+        refreshAccessTokenExpiresAt: new Date(Date.now() + data.refresh_token_expires_in * 1000),
+      },
+    };
+  }
+
+  /**
+   * Get the service app ID.
+   * @returns {string}
+   */
+  public getServiceAppId(): string {
+    return this.serviceAppId;
+  }
+
+  /**
+   * Get the service app authorization for a given organization ID.
+   * @param {string} orgId - The organization ID.
+   * @returns {Promise<ServiceAppAuthorization>}
+   */
+  public async getServiceAppAuthorization(orgId: string): Promise<ServiceAppAuthorization> {
+    if (!this.serviceAppAuthorizations[orgId]) {
+      return Promise.reject(new Error('Service app authorization not found'));
+    }
+
+    return Promise.resolve(this.serviceAppAuthorizations[orgId]);
+  }
+
+  /**
+   * Refresh the access token using the refresh token.
+   * @param {string} orgId - The organization ID.
+   * @returns {Promise<void>}
+   */
+  public async refreshServiceAppAccessToken(
+    orgId: string,
+    headers: Record<string, string> = {}
+  ): Promise<void> {
+    const serviceAppAuthorization = await this.getServiceAppAuthorization(orgId);
+    const refreshToken = serviceAppAuthorization?.token.refreshToken;
+    if (!refreshToken) {
+      throw new Error('Refresh token is undefined');
+    }
+    try {
+      const response: Response = await fetch(`${this.baseUrl}/access_token`, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded', // https://developer.webex.com/docs/login-with-webex#access-token-endpoint
+          ...headers,
+        },
+        body: new URLSearchParams({
+          grant_type: 'refresh_token',
+          client_id: this.clientId,
+          client_secret: this.clientSecret,
+          refresh_token: refreshToken,
+        }),
+      });
+
+      if (!response.ok) {
+        throw new Error(`Failed to refresh access token: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
+    } catch (error) {
+      console.error('Error refreshing access token:', error);
+      // Remove the serviceAppAuthorization token from here since we can't do anything more with it
+      // TODO: Have an event mechanism to notify the user of the error
+      delete this.serviceAppAuthorizations[orgId];
+      throw error;
+    }
+  }
+
+  /**
+   * Retrieve a new service app token after authorization.
+   * @param {string} orgId - The organization ID.
+   * @param {string} personalAccessToken - The service app owner's personal access token or token from an integration that has the scope `spark:applications_token`.
+   * @returns {Promise<void>}
+   */
+  public async retrieveTokenAfterAuthorization(
```

Comment:
Should we rename this to something like `getServiceAppTokenUsingPAT()`?

Reply:
Done